### PR TITLE
Ra 147

### PR DESCRIPTION
**Pull request by vguzmanp**
Should apply without many conflicts over RA-138 once it's merged

### CODE REVIEW

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 1]
```diff
@@ -209,6 +218,89 @@ void PropertyChangedItem (object sender, PropertyChangedEventArgs e)
...
+			List<TViewModel> selected = new List<TViewModel> ();
+
+			foreach (var path in Selection.GetSelectedRows()) {
+				Model.GetIterFromString (out iter, path.ToString ());
+				TViewModel selectedViewModel = Model.GetValue (iter, COL_DATA) as TViewModel;
+				if (selectedViewModel != null) {
+					selected.Add (selectedViewModel);
+				}
+			}
+			ViewModel.Selection.Replace (selected);
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: This should be done in ViewModel, something like **VieModel.SelectionReplace(selected);**

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 2]
```diff
@@ -209,6 +218,89 @@ void PropertyChangedItem (object sender, PropertyChangedEventArgs e)
...
+				/* Hack to make it actually resort */
+				sort.SetSortFunc (COL_DATA, HandleSort);
+			}
+
+			if (e.PropertyName == "FilterText") {
+				filter.Refilter ();
+			}
+
+			if (e.PropertyName == "Selection") {
+				//Sincronization of the first external selection
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Why is this needed?

File: VAS.UI.Gtk2/UI/Menu/MenuHelpers.cs [hunk 3]
```diff
@@ -64,36 +65,34 @@ static public void FillAddToPlaylistMenu (MenuItem addToPlaylistMenu, Project pr
...
 					);
 				};
-				plMenu.ShowAll ();
-				addToPlaylistMenu.Submenu = plMenu;
 			}
+			item = new MenuItem (Catalog.GetString ("Create new playlist..."));
+			plMenu.Append (item);
+			item.Activated += (sender, e) => {
+				IEnumerable<IPlaylistElement> elements = events.Select (p => new PlaylistPlayElement (p));
+				App.Current.EventsBroker.Publish (
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/Menu/MenuHelpers.cs]: This could be changed to:
item.PublishEvent (new AddPlaylistElementEvent {
	Playlist = null,
       PlaylistElements = elements.ToList ()
});

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 4]
```diff
@@ -35,14 +35,21 @@ namespace VAS.UI.Common
...
 	{
+		protected const int COL_DATA = 0;
 		protected TreeStore store;
 		TCollectionViewModel viewModel;
 		protected Dictionary<IViewModel, TreeIter> dictionaryStore;
 		protected Dictionary<INotifyCollectionChanged, TreeIter> dictionaryNestedParent;
 
+		protected TreeModelFilter filter;
+		protected TreeModelSort sort;
+
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Extra line

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 5]
```diff
@@ -33,12 +35,16 @@ namespace VAS.UI.Common
 	/// </summary>
 	public class TreeViewBase<TCollectionViewModel, TModel, TViewModel> : Gtk.TreeView, IView<TCollectionViewModel>
 		where TCollectionViewModel : CollectionViewModel<TModel, TViewModel>
-		where TViewModel: IViewModel<TModel>, new()
+		where TViewModel: class, IViewModel<TModel>, new()
```
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: There were some warnings about `TViewModel` possibly being a value type, and the warning itself recommended this.

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs (hunk 2 above)
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Honestly, no idea. It was copied from RAProjectTreeView

File: VAS.UI.Gtk2/UI/Common/PlaylistTreeView.cs [hunk 6]
```diff
@@ -15,15 +15,9 @@
 //  along with this program; if not, write to the Free Software
 //  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 //
-//  Copyright (C) 2016 Fluendo S.A.
-using System;
-using System.Collections.Generic;
-using System.Collections.ObjectModel;
-using System.Dynamic;
-using System.Linq;
+
```
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Common/PlaylistTreeView.cs]: I believe a line between the copyright comment and the using is a good thing, don't you?

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs (hunk 4 above)
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Fixed

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 7]
```diff
@@ -147,6 +155,76 @@ void PropertyChangedItem (object sender, PropertyChangedEventArgs e)
...
+			}
+		}
+
+		protected virtual int HandleSort (TreeModel model, TreeIter a, TreeIter b)
+		{
+			// FIXME: Implement a generic sort for all TViewModels
+			return 0;
+		}
+
+		protected virtual bool HandleFilter (Gtk.TreeModel model, Gtk.TreeIter iter)
```
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Fixed

File: VAS.Core/MVVMC/CollectionViewModel.cs [hunk 8]
```diff
@@ -43,8 +43,7 @@ public CollectionViewModel ()
 		{
 			Model = new ObservableCollection<TModel> ();
 			ViewModels = new ObservableCollection<TViewModel> ();
-			Selection = new RangeObservableCollection<TViewModel> ();
-			Selection.CollectionChanged += HandleSelectionChanged;
+
```
Comment (vguzmanp) [inline VAS.Core/MVVMC/CollectionViewModel.cs]: Fixed
